Support all babel transform options

tquetano-netflix opened the pull request:
## Reason for change

The top-level `presets` API is inherently limiting, as only Babel presets can be used to provide a babel configuration for the plugin. Ideally all the babel options would be supported, as is called out in #28 .

## Change

Modify API to have `babelTransformOptions` be the entrypoint for babel options that can be overridden, and that is merged into the options created by `getBabelOpts`.

### Note

Because of some internal dependencies, I was not able to verify everything locally (out of the box, 22 of 31 unit tests fail locally). My changes do not cause any additional tests to fail, and it passes typecheck, so it _should_ be okay.

I'm trying to figure out how to do a simple unit test to verify custom options output, but there are a few layers to peel back there. Advice would be welcome.

NickHeiner commented on src/run-codemod-on-file.ts:
```diff
@@ -138,20 +138,21 @@ export default async function runCodemodOnFile(
         const babelOpts = {
           ...getBabelOpts(),
-          ..._.pick(codemod, 'presets'),
+          ..._.pick(codemod, 'babelTransformOptions'),
```
It seems like `getBabelOpts` should have higher precedence than `babelTransformOptions`, right? If the user passes `{ast: false}`, it'll break the codemod runner.

Generally, I'm a bit concerned this feature will be tricky to use, because it's hard for the user to know what they can safely pass, and what will interfere with what the codemod runner needs. But I'm ok just saying in the docs that this is "advanced", and the only thing you can safely pass is "presets". :smile:

What do you think?

NickHeiner replied:
nm, I see you basically responded to this same question [earlier](https://github.com/NickHeiner/jscodemod/issues/28#issuecomment-923047678).

NickHeiner commented on src/run-codemod-on-file.ts:
```diff
@@ -138,20 +138,21 @@ export default async function runCodemodOnFile(
         const babelOpts = {
           ...getBabelOpts(),
-          ..._.pick(codemod, 'presets'),
+          ..._.pick(codemod, 'babelTransformOptions'),
```
This isn't quite what we want. `_.pick(a, 'b')` results in `{b}`, so our spread is:

```js
{
  ...getBabelOpts(),
  ...{babelTransformOptions}  
}
```

We just want `...codemod.babelTransformOptions`.

I made this fix on my local branch.

tquetano-netflix replied:
Yeah my lodash is rusty. I don't understand why `_.pick` is needed here at all, vs just:

```ts
...getBabelOpts(),
presets: codemod.presets,
```

But I was trying to align with conventions of the project.

NickHeiner commented on src/run-codemod-on-file.ts:
```diff
@@ -138,20 +138,21 @@ export default async function runCodemodOnFile(
-           * https://github.com/benjamn/recast/issues/834
-           */
           parserOpts: {
+            ..._.pick(codemod, 'babelTransformOptions.parserOpts'),
```
`pick` uses the second argument as a key – it doesn't do path parsing. You'd want `..._.pick(codemod. babelTransformOptions, 'parserOpts')`. 

I fixed this object spreading on my local branch.

NickHeiner commented on __tests__/index.ts:
```diff
@@ -316,6 +316,16 @@ describe('happy path', () => {
+  createTest({
+    testName: 'getPlugin supports custom `babelTransformOptions`',
+    fixtureName: 'custom-babel-options',
+    spawnArgs: ['--codemod', 'codemod.js', 'source'],
```
This needs to be `codemod.ts`.

NickHeiner replied:
I've made this fix locally.

tquetano-netflix commented:
I've updated the PR with the changes mentioned, and also updated `getBabelOpts` to have a more bespoke merge setup. I can write a unit test for the custom `plugins` merging, but I wanted feedback on the general approach before I went that direction.
